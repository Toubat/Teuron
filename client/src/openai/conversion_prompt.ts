import { ARROW, VAR_TYPE_VALUE_STOP_SYMBOL } from "../shared";
import {
  createVarTypeDeclerationString,
  createVarTypeValueString,
  VarType,
  VarTypeDecleration,
  VarTypeValue,
} from "./var_types";

interface CraeteConversionExample<T extends VarType> {
  prompt: string;
  output: VarTypeValue<T>;
}

interface CreateConversionOptions<T extends VarType> {
  typeNode: VarTypeDecleration<T>;
  prompt: string;
  functionName: string;
  returnName?: string;
  promptDescription?: string;
  resultDescription?: string;
  examples?: CraeteConversionExample<T>[];
}

export function createArgumentDescription(
  paramName: string,
  varType: string,
  description: string
): string {
  return `- ${paramName} (${varType}): ${description}`;
}

export function createConversionPrompt<T extends VarType>({
  typeNode,
  prompt,
  functionName,
  returnName = "ReturnType",
  promptDescription = "A natural language description",
  resultDescription = "The result of the function",
  examples = [],
}: CreateConversionOptions<T>): string {
  const prompts: string[] = [];
  const push = (str: string) => prompts.push(str);

  push(`type ${returnName} = ${createVarTypeDeclerationString(typeNode)};\n`);
  push(`/*`);
  push(`Input:`);
  push(` ${createArgumentDescription("prompt", VarType.String, promptDescription)}`);
  push(`Return:`);
  push(` ${createArgumentDescription("result", returnName, resultDescription)}`);

  if (examples.length > 0) {
    push(`Examples:`);
    examples.forEach(({ prompt, output }) => {
      push(`${functionName}("${prompt}")`);
      push(`${ARROW} ${createVarTypeValueString(output)}${VAR_TYPE_VALUE_STOP_SYMBOL}`);
    });
  }

  push(`*/\n`);
  push(`function ${functionName}(prompt: string): ${returnName};\n`);
  push(
    `// Given input of ${functionName}, return the output (end with ${VAR_TYPE_VALUE_STOP_SYMBOL})`
  );
  push(`${functionName}("${prompt}")`);
  push(`${ARROW} `);

  return prompts.join("\n");
}
