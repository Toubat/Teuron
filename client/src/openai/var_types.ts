import { isArray, isBoolean, isNumber, isObject, isString } from "../shared";

export enum VarType {
  Boolean = "boolean",
  Number = "number",
  String = "string",
  Array = "array",
  Object = "object",
}

export type VarTypeDecleration<T = VarType> = {
  type: T;
  inner: VarTypeDeclerationInner<T>;
};

export type VarTypeDeclerationInner<T = VarType> = T extends VarType.Boolean
  ? VarType.Boolean
  : T extends VarType.Number
  ? VarType.Number
  : T extends VarType.String
  ? VarType.String
  : T extends VarType.Object
  ? Record<string, VarTypeDecleration>
  : T extends VarType.Array
  ? VarTypeDecleration
  : never;

export type VarTypeValue<T = VarType> = T extends VarType.Boolean
  ? boolean
  : T extends VarType.Number
  ? number
  : T extends VarType.String
  ? string
  : T extends VarType.Object
  ? {
      [key: string]: VarTypeValue;
    }
  : T extends VarType.Array
  ? VarTypeValue[]
  : never;

export function createVarTypeDeclerationString<T extends VarType>(
  typeNode: VarTypeDecleration<T>,
  indent = 1
): string {
  const { type, inner } = typeNode;
  const indentStr = "\t".repeat(indent);

  switch (type) {
    case VarType.Boolean:
      return VarType.Boolean;

    case VarType.Number:
      return VarType.Number;

    case VarType.String:
      return VarType.String;

    case VarType.Array:
      return `${createVarTypeDeclerationString(inner as VarTypeDecleration, indent)}[]`;

    case VarType.Object:
      const lines: string[] = [];
      const push = (str: string) => lines.push(str);

      push("{");
      Object.entries(inner as Record<string, VarTypeDecleration>).forEach(([key, value]) => {
        push(`${indentStr}"${key}": ${createVarTypeDeclerationString(value, indent + 1)};`);
      });
      push(`${indentStr.slice(1)}}`);
      return lines.join("\n");

    default:
      throw new Error("Invalid type");
  }
}

export function createVarTypeValueString<T extends VarType>(value: VarTypeValue<T>): string {
  let raw: string;

  if (isBoolean(value)) {
    raw = value ? "true" : "false";
  } else if (isNumber(value)) {
    raw = value.toString();
  } else if (isString(value)) {
    raw = `"${value}"`;
  } else if (isArray(value)) {
    raw = `[${(value as VarTypeValue[]).map((item) => createVarTypeValueString(item)).join(", ")}]`;
  } else if (isObject(value)) {
    raw = `{${Object.entries(value as Record<string, VarTypeValue>)
      .map(([key, value]) => `"${key}": ${createVarTypeValueString(value)}`)
      .join(", ")}}`;
  } else {
    throw new Error("Invalid value");
  }

  return raw;
}
