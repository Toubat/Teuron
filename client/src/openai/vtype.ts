import { isArray, isBoolean, isNumber, isObject, isString } from "../shared";

export enum VType {
  Boolean = "boolean",
  Number = "number",
  String = "string",
  Array = "array",
  Object = "object",
}

type BoolDecleration = {
  type: VType.Boolean;
  inner: VType.Boolean;
};

type NumberDecleration = {
  type: VType.Number;
  inner: VType.Number;
};

type StringDecleration = {
  type: VType.String;
  inner: VType.String;
};

type ArrayDecleration = {
  type: VType.Array;
  inner: VTypeDecleration;
};

type ObjectDecleration = {
  type: VType.Object;
  inner: Record<string, VTypeDecleration>;
};

export type VTypeDecleration<T extends VType = any> = T extends VType.Boolean
  ? BoolDecleration
  : T extends VType.Number
  ? NumberDecleration
  : T extends VType.String
  ? StringDecleration
  : T extends VType.Array
  ? ArrayDecleration
  : T extends VType.Object
  ? ObjectDecleration
  : never;

export type VTypeValue<T extends VType = any> = T extends VType.Boolean
  ? boolean
  : T extends VType.Number
  ? number
  : T extends VType.String
  ? string
  : T extends VType.Object
  ? Record<string, VTypeValue>
  : T extends VType.Array
  ? VTypeValue[]
  : never;

export function createVTypeDeclerationString<T extends VType>(
  typeNode: VTypeDecleration<T>,
  indent = 1
): string {
  const { type, inner } = typeNode;
  const indentStr = "\t".repeat(indent);

  switch (type) {
    case VType.Boolean:
    case VType.Number:
    case VType.String:
      return inner;

    case VType.Array:
      return `${createVTypeDeclerationString(inner, indent)}[]`;

    case VType.Object:
      const lines: string[] = [];
      const push = (str: string) => lines.push(str);

      push("{");
      Object.entries(inner).forEach(([key, value]) => {
        push(`${indentStr}"${key}": ${createVTypeDeclerationString(value, indent + 1)};`);
      });
      push(`${indentStr.slice(1)}}`);
      return lines.join("\n");

    default:
      throw new Error("Invalid type");
  }
}

export function createVTypeValueString<T extends VType>(value: VTypeValue<T>): string {
  let raw: string;

  if (isBoolean(value)) {
    raw = value ? "true" : "false";
  } else if (isNumber(value)) {
    raw = value.toString();
  } else if (isString(value)) {
    raw = `"${value}"`;
  } else if (isArray(value)) {
    raw = `[${(value as VTypeValue[]).map((item) => createVTypeValueString(item)).join(", ")}]`;
  } else if (isObject(value)) {
    raw = `{${Object.entries(value as Record<string, VTypeValue>)
      .map(([key, value]) => `"${key}": ${createVTypeValueString(value)}`)
      .join(", ")}}`;
  } else {
    throw new Error("Invalid value");
  }

  return raw;
}
