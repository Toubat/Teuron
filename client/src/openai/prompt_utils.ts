import { VType, VTypeValue } from "./vtype";

export interface PromptExample<T extends VType> {
  prompt: string;
  output: VTypeValue<T>;
}

export function createArgumentDescription(
  paramName: string,
  varType: string,
  description: string
): string {
  return `- ${paramName} (${varType}): ${description}`;
}
