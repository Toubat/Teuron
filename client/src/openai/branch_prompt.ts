import { EMPTY_STRING, FALSE, TRUE } from "../shared";
import { createInput, createOutput, PromptExample } from "./prompt_utils";
import { VType } from "./vtype";

const BRANCH_PROMPT_INTRO = `Given a context and a hypothesis, return whether the hypothesis is true or false based on the context.\n`;

interface BranchOptions {
  context: string;
  hypothesis: string;
  examples?: PromptExample<VType.Boolean>[];
}

export function createBranchPrompt({ context, hypothesis, examples = [] }: BranchOptions): string {
  const prompts: string[] = [];
  const push = (str: string) => prompts.push(str);

  push(BRANCH_PROMPT_INTRO);

  examples.forEach((example) => {
    push(createInput(example.prompt));
    push(createOutput(`${example.output ? TRUE : FALSE}\n`));
  });

  push(createInput(`Context: ${context}, Hypothesis: ${hypothesis}`));
  push(createOutput(EMPTY_STRING));

  return prompts.join("\n");
}
