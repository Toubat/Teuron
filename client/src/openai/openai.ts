import { Configuration, OpenAIApi } from "openai";
import type { CreateCompletionRequest } from "openai";

// TODO: delete this
const API_KEY = "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";
const configuration = new Configuration({
  apiKey: API_KEY,
});
const openai = new OpenAIApi(configuration);

export async function createCompletion(request: CreateCompletionRequest) {
  const completion = await openai.createCompletion(request);
  return completion;
}
