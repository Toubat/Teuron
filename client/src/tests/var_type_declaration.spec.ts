import { describe, it, expect } from "vitest";
import type { VarTypeDecleration } from "../openai/var_types";
import { VarType, createVarTypeDeclerationString } from "../openai/var_types";

describe("createVarTypeDeclerationString", () => {
  it("number type", () => {
    const typeNode: VarTypeDecleration<VarType.Number> = {
      type: VarType.Number,
      inner: VarType.Number,
    };

    const result = createVarTypeDeclerationString(typeNode);
    expect(result).toBe("number");
  });

  it("boolean type", () => {
    const typeNode: VarTypeDecleration<VarType.Boolean> = {
      type: VarType.Boolean,
      inner: VarType.Boolean,
    };

    const result = createVarTypeDeclerationString(typeNode);
    expect(result).toBe("boolean");
  });

  it("string type", () => {
    const typeNode: VarTypeDecleration<VarType.String> = {
      type: VarType.String,
      inner: VarType.String,
    };

    const result = createVarTypeDeclerationString(typeNode);
    expect(result).toBe("string");
  });

  it("simple array type", () => {
    const typeNode: VarTypeDecleration<VarType.Array> = {
      type: VarType.Array,
      inner: {
        type: VarType.String,
        inner: VarType.String,
      },
    };

    const result = createVarTypeDeclerationString(typeNode);
    expect(result).toBe("string[]");
  });

  it("simple object type", () => {
    const typeNode: VarTypeDecleration<VarType.Object> = {
      type: VarType.Object,
      inner: {
        a: {
          type: VarType.String,
          inner: VarType.String,
        },
        b: {
          type: VarType.Number,
          inner: VarType.Number,
        },
      },
    };

    const result = createVarTypeDeclerationString(typeNode);
    expect(result).toMatchSnapshot();
  });

  it("array of simple object", () => {
    const arrayOfObject: VarTypeDecleration<VarType.Array> = {
      type: VarType.Array,
      inner: {
        type: VarType.Object,
        inner: {
          a: {
            type: VarType.String,
            inner: VarType.String,
          },
          b: {
            type: VarType.Number,
            inner: VarType.Number,
          },
        },
      },
    };

    const result = createVarTypeDeclerationString(arrayOfObject);
    expect(result).toMatchSnapshot();
  });

  it("object of array of simple object", () => {
    const arrayOfObject: VarTypeDecleration<VarType.Array> = {
      type: VarType.Array,
      inner: {
        type: VarType.Object,
        inner: {
          a: {
            type: VarType.String,
            inner: VarType.String,
          },
          b: {
            type: VarType.Number,
            inner: VarType.Number,
          },
        },
      },
    };

    const objectOfArrayOfObject: VarTypeDecleration<VarType.Object> = {
      type: VarType.Object,
      inner: {
        a: {
          type: VarType.Number,
          inner: VarType.Number,
        },
        b: arrayOfObject,
      },
    };

    const result = createVarTypeDeclerationString(objectOfArrayOfObject);
    expect(result).toMatchSnapshot();
  });

  it("array of object of array of simple object", () => {
    const arrayOfObject: VarTypeDecleration<VarType.Array> = {
      type: VarType.Array,
      inner: {
        type: VarType.Object,
        inner: {
          a: {
            type: VarType.String,
            inner: VarType.String,
          },
          b: {
            type: VarType.Number,
            inner: VarType.Number,
          },
        },
      },
    };

    const objectOfArrayOfObject: VarTypeDecleration<VarType.Object> = {
      type: VarType.Object,
      inner: {
        a: {
          type: VarType.Number,
          inner: VarType.Number,
        },
        b: arrayOfObject,
      },
    };

    const arrayOfObjectOfArrayOfObject: VarTypeDecleration<VarType.Array> = {
      type: VarType.Array,
      inner: objectOfArrayOfObject,
    };

    const result = createVarTypeDeclerationString(arrayOfObjectOfArrayOfObject);
    expect(result).toMatchSnapshot();
  });

  it("complex object", () => {
    const complexObject: VarTypeDecleration<VarType.Object> = {
      type: VarType.Object,
      inner: {
        a: {
          type: VarType.String,
          inner: VarType.String,
        },
        b: {
          type: VarType.Number,
          inner: VarType.Number,
        },
        c: {
          type: VarType.Boolean,
          inner: VarType.Boolean,
        },
        d: {
          type: VarType.Array,
          inner: {
            type: VarType.String,
            inner: VarType.String,
          },
        },
        e: {
          type: VarType.Array,
          inner: {
            type: VarType.Number,
            inner: VarType.Number,
          },
        },
        f: {
          type: VarType.Array,
          inner: {
            type: VarType.Boolean,
            inner: VarType.Boolean,
          },
        },
        g: {
          type: VarType.Array,
          inner: {
            type: VarType.Object,
            inner: {
              a: {
                type: VarType.String,
                inner: VarType.String,
              },
              b: {
                type: VarType.Number,
                inner: VarType.Number,
              },
              c: {
                type: VarType.Array,
                inner: {
                  type: VarType.Object,
                  inner: {
                    a: {
                      type: VarType.String,
                      inner: VarType.String,
                    },
                    b: {
                      type: VarType.Number,
                      inner: VarType.Number,
                    },
                  },
                },
              },
            },
          },
        },
      },
    };

    const result = createVarTypeDeclerationString(complexObject);
    expect(result).toMatchSnapshot();
  });

  it("complex nested object & array", () => {
    const complexObject: VarTypeDecleration<VarType.Object> = {
      type: VarType.Object,
      inner: {
        arr: {
          type: VarType.Array,
          inner: {
            type: VarType.Object,
            inner: {
              a: {
                type: VarType.String,
                inner: VarType.String,
              },
              b: {
                type: VarType.Array,
                inner: {
                  type: VarType.Object,
                  inner: {
                    a: {
                      type: VarType.Object,
                      inner: {
                        a: {
                          type: VarType.Object,
                          inner: {
                            a: {
                              type: VarType.Number,
                              inner: VarType.Number,
                            },
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          },
        },
      },
    };

    const result = createVarTypeDeclerationString(complexObject);
    expect(result).toMatchSnapshot();
  });
});
