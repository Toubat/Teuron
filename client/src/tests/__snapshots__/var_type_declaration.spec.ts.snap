// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createVarTypeDeclerationString > array of object of array of simple object 1`] = `
"{
	\\"a\\": number;
	\\"b\\": {
		\\"a\\": string;
		\\"b\\": number;
	}[];
}[]"
`;

exports[`createVarTypeDeclerationString > array of simple object 1`] = `
"{
	\\"a\\": string;
	\\"b\\": number;
}[]"
`;

exports[`createVarTypeDeclerationString > complex nested object & array 1`] = `
"{
	\\"arr\\": {
		\\"a\\": string;
		\\"b\\": {
			\\"a\\": {
				\\"a\\": {
					\\"a\\": number;
				};
			};
		}[];
	}[];
}"
`;

exports[`createVarTypeDeclerationString > complex object 1`] = `
"{
	\\"a\\": string;
	\\"b\\": number;
	\\"c\\": boolean;
	\\"d\\": string[];
	\\"e\\": number[];
	\\"f\\": boolean[];
	\\"g\\": {
		\\"a\\": string;
		\\"b\\": number;
		\\"c\\": {
			\\"a\\": string;
			\\"b\\": number;
		}[];
	}[];
}"
`;

exports[`createVarTypeDeclerationString > object of array of simple object 1`] = `
"{
	\\"a\\": number;
	\\"b\\": {
		\\"a\\": string;
		\\"b\\": number;
	}[];
}"
`;

exports[`createVarTypeDeclerationString > simple object type 1`] = `
"{
	\\"a\\": string;
	\\"b\\": number;
}"
`;

exports[`var_type > array of object of array of simple object 1`] = `
"{
	\\"a\\": number;
	\\"b\\": {
		\\"a\\": string;
		\\"b\\": number;
	}[];
}[]"
`;

exports[`var_type > array of simple object 1`] = `
"{
	\\"a\\": string;
	\\"b\\": number;
}[]"
`;

exports[`var_type > complex nested object & array 1`] = `
"{
	\\"arr\\": {
		\\"a\\": string;
		\\"b\\": {
			\\"a\\": {
				\\"a\\": {
					\\"a\\": number;
				};
			};
		}[];
	}[];
}"
`;

exports[`var_type > complex object 1`] = `
"{
	\\"a\\": string;
	\\"b\\": number;
	\\"c\\": boolean;
	\\"d\\": string[];
	\\"e\\": number[];
	\\"f\\": boolean[];
	\\"g\\": {
		\\"a\\": string;
		\\"b\\": number;
		\\"c\\": {
			\\"a\\": string;
			\\"b\\": number;
		}[];
	}[];
}"
`;

exports[`var_type > object of array of simple object 1`] = `
"{
	\\"a\\": number;
	\\"b\\": {
		\\"a\\": string;
		\\"b\\": number;
	}[];
}"
`;

exports[`var_type > simple object type 1`] = `
"{
	\\"a\\": string;
	\\"b\\": number;
}"
`;

exports[`var_type_declaration > array of object of array of simple object 1`] = `
"{
	\\"a\\": number;
	\\"b\\": {
		\\"a\\": string;
		\\"b\\": number;
	}[];
}[]"
`;

exports[`var_type_declaration > array of simple object 1`] = `
"{
	\\"a\\": string;
	\\"b\\": number;
}[]"
`;

exports[`var_type_declaration > complex nested object & array 1`] = `
"{
	\\"arr\\": {
		\\"a\\": string;
		\\"b\\": {
			\\"a\\": {
				\\"a\\": {
					\\"a\\": number;
				};
			};
		}[];
	}[];
}"
`;

exports[`var_type_declaration > complex object 1`] = `
"{
	\\"a\\": string;
	\\"b\\": number;
	\\"c\\": boolean;
	\\"d\\": string[];
	\\"e\\": number[];
	\\"f\\": boolean[];
	\\"g\\": {
		\\"a\\": string;
		\\"b\\": number;
		\\"c\\": {
			\\"a\\": string;
			\\"b\\": number;
		}[];
	}[];
}"
`;

exports[`var_type_declaration > object of array of simple object 1`] = `
"{
	\\"a\\": number;
	\\"b\\": {
		\\"a\\": string;
		\\"b\\": number;
	}[];
}"
`;

exports[`var_type_declaration > simple object type 1`] = `
"{
	\\"a\\": string;
	\\"b\\": number;
}"
`;
